src/lib/protocol/arp.lua

```diff
@@ -0,0 +1,105 @@
+module(..., package.seeall)
+
+-- RFC 826 (ARP) requests and replies.
+-- Note: all incoming configurations are assumed to be in network byte order.
+
+local ethernet = require("lib.protocol.ethernet")
+local header = require("lib.protocol.header")
+local packet = require("core.packet")
+
+local ffi = require("ffi")
+local C = ffi.C
+
+--[[ Packet format (IPv4/Ethernet), as described on Wikipedia.
+
+Internet Protocol (IPv4) over Ethernet ARP packet
+octet offset 	0 	1
+0 	Hardware type (HTYPE)
+2 	Protocol type (PTYPE)
+4 	Hardware address length (HLEN) 	Protocol address length (PLEN)
+6 	Operation (OPER)
+8 	Sender hardware address (SHA) (first 2 bytes)
+10 	(next 2 bytes)
+12 	(last 2 bytes)
+14 	Sender protocol address (SPA) (first 2 bytes)
+16 	(last 2 bytes)
+18 	Target hardware address (THA) (first 2 bytes)
+20 	(next 2 bytes)
+22 	(last 2 bytes)
+24 	Target protocol address (TPA) (first 2 bytes)
+26 	(last 2 bytes)
+--]]
+
+local arp_header_t = ffi.typeof[[
+   struct {
+   uint16_t arp_htype;
+   uint16_t arp_ptype;
+   uint8_t arp_hlen;
+   uint8_t arp_plen;
+   uint16_t arp_oper;
+   uint8_t arp_sha[6];
+   uint8_t arp_spa[4];
+   uint8_t arp_tha[6];
+   uint8_t arp_tpa[4];
+   } __attribute__((packed))
+]]
+
+local arp = subClass(header)
+-- ARP class variables
+arp._name = "arp"
+arp._header_type = arp_header_t
+arp._header_ptr_type = ffi.typeof("$*", arp_header_t)
+arp._ulp = { method = nil }
+
+-- Constants
+local arp_request = C.ntohs(1)
+local arp_reply = C.ntohs(2)
+
+local unknown_eth = ethernet:pton("00:00:00:00:00:00")
```

Comment:
The format of an ethernet addresses is something I personally don't pay attention to but I recall one Snabb Switch bug that reported poor performance on a test and the reason was that the chosen destination MAC address started with 01, which means multicast.

The Ethernet protocol [1] states that the least significant bit and second least significant bit of the most significant byte of a MAC address have special meanings:

- b1: if 0, address is unicast. if 1, address is multicast.
- b2: if 0, globally unique. if 1, locally administered.

To avoid potential problems, I think it's better to use an unicast local administered address. There are 4 family of addresses like that in the MAC address range: X2-XX-XX, X6-XX-XX, XA-XX-XX and XE-XX-XX. For instance, 02:00:00:00:00:00 would work.

On the other hand, according to [2] addresses starting with 00:00:00 are reserved. In addition to that, googling and all zero MAC addresses reports issues on a model of Cisco switches that discard incoming frames with an all zero address as source address [3].

[1] https://en.wikipedia.org/wiki/MAC_address
[1] http://www.iana.org/assignments/ethernet-numbers/ethernet-numbers.xhtml.
[2] https://supportforums.cisco.com/discussion/10895666/invalid-source-mac-address-000000000000